Optimize array pattern matching by caching #deconstruct value

palkan opened the pull request:
This is follow-up for [the discussion](https://github.com/ruby/ruby/pull/2948#issuecomment-609034219) started in #2948.

## Overview

We can improve array patterns matching performance by keeping the result of the `#deconstruct`  call (or the fact, that the value doesn't respond to it).

The solution proposed in this PR is the following:
- In `compile_case3`:
  - Reserve a stack slot for `#deconstruct` value (with the default value `nil` — haven't tried to deconstruct)
- In `iseq_compile_pattern_each`:
  - Added `deconstructed_pos` argument which is used to track the _distance_ to the reserved slot (since it could be different depending on the node type).

## Benchmark

Compared with the current master ([benchmark](https://github.com/ruby/ruby/commit/ae3ab8ed3a3bb47231eb943d2b4a3168dca771b2)):

```sh
                    first_match
              ./ruby:   1137096.7 i/s 
  /etc/ruby/bin/ruby:   1112537.0 i/s - 1.02x  slower

                     second_match
              ./ruby:    995800.6 i/s 
  /etc/ruby/bin/ruby:    961526.1 i/s - 1.04x  slower

                      third_match
              ./ruby:    895467.2 i/s 
  /etc/ruby/bin/ruby:    822866.8 i/s - 1.09x  slower

                     fourth_match
              ./ruby:    780673.0 i/s 
  /etc/ruby/bin/ruby:    693164.3 i/s - 1.13x  slower
```

## Limitations & future ideas

Currently, only the top-level array patterns are affected. Adding a similar optimization to nested patterns would require additional (and, I think, in most cases useless) stack allocations. Thus, this approach is not optimal, still an ad-hoc optimization.

The proper solution would be to analyze the pattern tree first and generate an optimal _execution path_, which wouldn't have duplicate structural checks. I plan to work on this too, but that would require much more time than the proposed minor improvement. So, I think this patch is still worth to be considered.

palkan commented:
@k-tsj Could you please take a look?

k-tsj commented:
Great work.

I am positive to merge this pull request.
However, I have been very busy lately, so please give me some time to review.

Could you add some tests?
For example, it seems we need following case.

```ruby
case [0]
in []
in [1] | [0]
  true 
end
```

palkan commented:
@k-tsj Added some tests, found some bugs and fixed them 🙂

Unfortunately, the initial idea of including deconstruct instructions only in the first pattern doesn't work well with constant matching, since it's checked first.
So, I rewrote this logic by relying only on the current reserved value on the stack and by adding a couple of new labels to the compiled instructions for branching. The current logic could described as follows:
- Check reserved value:
  - if it's `nil` — we haven't tried to deconstruct yet, jump to `deconstruct`
  - if it's `false` — the matchee doesn't respond to `#deconsturct`, jump to `match_failed`
  - if it's non-falsey — jump to `deconstructed`.

- In the `deconstruct` state:
  - if doesn't respond to `#deconstruct` — put `false` on the stack (instead of `nil`)
  - if does — try to deconstruct and put the result on the stack.

k-tsj commented:
I found a bug.
Could you check it?

```
$ ruby -e 'case [[0]]; in [[*a]]; p a;end'
(Expected)
[0]

(Actual)
-e:1:in `<main>': undefined method `length' for 0:Integer (NoMethodError)
```

palkan commented:
@k-tsj Fixed (and added a few more tests).

k-tsj commented:
Could you rebase to master?

palkan commented:
@k-tsj Rebased and extracted this functionality into `iseq_compile_array_deconstruct` to re-use in array and find patterns.

k-tsj commented:
Oops, it seems that changes in 37cd877dbd70 cause conflict.
After it's fixed, I'll merge this PR.

palkan commented:
@k-tsj Fixed!

k-tsj commented:
Thanks!
I've merged.
